New landing page design

shravanngoswamii opened the pull request:
Continuing https://github.com/TuringLang/turinglang.github.io/pull/105
Preview the changes: https://turinglang.org/pr-previews/119

shravanngoswamii commented:
If anyone is interested, then please feel free to take on this PR!

simonsteiger commented:
I will definitely have time to finish this in July. Which points would you like to see addressed before this PR can be merged?

shravanngoswamii commented:
> Revisiting this. Is there a reason why the old PR was closed?

Apologies for the confusion. I don’t have admin permissions on this repo, and the previous PR had complex merge conflicts. Since it was from Simon’s fork, and I was adding some features to the landing page, I accidentally messed up some git operations. So I took those changes and opened a new PR from a branch within this repo to make it easier to manage.

> Honestly, it's not like the existing front page has any resources, so even a fairly clean slate would be a huge improvement over the current one.

Totally agree. However, before merging, I think we should either:
- Complete the light/dark mode feature and update that in the `docs` repo for consistency, or
- Move the color theme changes to a separate PR if we're not ready to finalize them yet.

shravanngoswamii commented:
Hi @penelopeysm, I’ve removed @simonsteiger’s styles/theme update from this PR and moved those changes to the current theme format. I also made a few updates to the theme and added some improvements to the news slider animation.

The full SCSS-based theme Simon added was actually a much better approach than the current one, but it needed some updates. For now, let’s focus on completing the landing page. I’ve backed up Simon’s theme, and I’ll open a separate PR for those styles once we’re done.

shravanngoswamii commented:
> I think I am still of the opinion that we don't need to add more _content_ right now.

Agree, I have commented everything from `Algorithms` section and added a Citation at the end, just take a look at it, let me know if you have suggestion on improving that citation part + I think we have a new paper on Turing, should we add that in citation or the old one I found on [main Turing.jl repo](https://github.com/TuringLang/Turing.jl/blob/main/CITATION.bib)

penelopeysm commented:
Very nice @shravanngoswamii! Could you also add this paper https://dl.acm.org/doi/10.1145/3711897?

I'll work on the placeholder text now.

penelopeysm commented:
There are some annoyances with the light/dark mode toggle (e.g. how it resets the animation), but I don't think they can be fixed while we are still using Quarto. (#110, but who has the time for that...)

yebai commented:
I suggest that we keep only these sections below. The other sections are good, but too detailed. The styling can be tweaked so all the sections are consistent (e.g. title centred or left aligned). 



<img width="1464" alt="image" src="https://github.com/user-attachments/assets/cfc1716a-b864-42db-a2d4-5fe099b34e0a" />

<img width="1224" alt="image" src="https://github.com/user-attachments/assets/dd4c1151-a777-4ff3-9ab6-2cf66b54fad7" />

yebai commented:
Also, let's look at the Alan Turing Institute website and try to add the ATI logo or branding material on our landing page. 

https://www.turing.ac.uk/

Cc @mhauru, who is coordinating this, I believe.

penelopeysm reviewed (commented):
Right. I'm done here.

Final comments:

1. Is there a way to scroll the news carousel with the mouse? Right now there's some dragging behaviour but I think just plain x-scrolling would be more intuitive.

2. The animation at the top replays when anything happens -- light/dark mode toggle is one but it also restarts when the window is resized. Is there a way to fix this?

shravanngoswamii commented:
> 1. Is there a way to scroll the news carousel with the mouse? Right now there's some dragging behaviour but I think just plain x-scrolling would be more intuitive.

Yep, I will do this, I was thinking of some better ways for this but let's try a mouse scroll



> There are some annoyances with the light/dark mode toggle (e.g. how it resets the animation), but I don't think they can be fixed while we are still using Quarto. (#110, but who has the time for that...)

> 2. The animation at the top replays when anything happens -- light/dark mode toggle is one but it also restarts when the window is resized. Is there a way to fix this?

Actually, this isn't a Quarto issue, something on our side is triggering this, let me have a look at it!

shravanngoswamii commented:
Here is the News Carousel code for landing page, putting it here for just in case we want to use it again in future:

<details><summary>News Carousel Code</summary>
<p>

```
Directory structure:
└── news/
    ├── _news-carousel.ejs
    └── news.qmd
```

#### File: _news-carousel.ejs
```qmd
/```{=html}
<div id="news-carousel-container">
    <div id="news-carousel-track">
        <% for (const item of items) { %>
            <div class="news-carousel-slide">
                <a href="<%- item.path %>" class="news-carousel-card">
                    <div class="news-carousel-card-body">
                        <h5 class="news-carousel-title"><%- item.title %></h5>
                        <p class="news-carousel-reading-time"><%- item['reading-time'] %> read</p>
                        <% if (item.description) { %>
                        <p class="news-carousel-description"><%- item.description %></p>
                        <% } %>
                        <div class="news-carousel-attribution">
                            <span class="news-carousel-author"><%- item.author %></span>
                            <span class="news-carousel-date"><%- item.date %></span>
                        </div>
                    </div>
                </a>
            </div>
        <% } %>
    </div>
</div>
/```
```


#### File: news.qmd
```qmd
---
listing:
  - id: news-carousel
    contents: 
      - "news/posts/*/index.qmd"
    template: _includes/news/_news-carousel.ejs
    sort: "date desc"
---

### News

::: {#news-carousel}
:::

\```{=html}
<div class="d-flex flex-column align-items-center gap-0">
  <a href="news/" class="button btn">
    See all news &rarr;
  </a>
</div>

<style>
/* Scoped styles for the news carousel to avoid conflicts with global themes */
#news-carousel-container {
  overflow: hidden;
  position: relative;
  width: 100%;
  cursor: grab;
}

#news-carousel-container.grabbing {
  cursor: grabbing;
}

#news-carousel-container #news-carousel-track {
  display: flex;
  align-items: stretch;
}

#news-carousel-container .news-carousel-slide {
  flex-shrink: 0;
  width: 33.3333%;
  padding: 0.5rem;
  box-sizing: border-box;
}

@media (max-width: 1024px) {
  #news-carousel-container .news-carousel-slide {
    width: 50%;
  }
}

@media (max-width: 768px) {
  #news-carousel-container .news-carousel-slide {
    width: 100%;
  }
}

#news-carousel-container .news-carousel-card {
  overflow: hidden;
  display: flex;
  flex-direction: column;
  height: 100%;
  text-decoration: none;
  border-radius: 0.5rem;
  transition: background-color 0.3s ease, border-color 0.3s ease;
  background-color: #ffffff;
  border: 1px solid #e9ecef;
  color: #212529;
}

#news-carousel-container .news-carousel-card:hover {
  background-color: #e9ecef;
  border-color: #dee2e6;
}

#news-carousel-container .news-carousel-card-body {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  padding: 1rem;
}

#news-carousel-container .news-carousel-title {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  margin-bottom: 0.25rem;
  color: #212529;
  font-weight: 700;
}

#news-carousel-container .news-carousel-reading-time,
#news-carousel-container .news-carousel-description,
#news-carousel-container .news-carousel-attribution {
  color: #6c757d;
}

#news-carousel-container .news-carousel-reading-time {
  font-size: 0.9em;
  margin-bottom: 0.75rem;
}

#news-carousel-container .news-carousel-description {
  flex-grow: 1;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
  line-clamp: 2;
  overflow: hidden;
  text-overflow: ellipsis;
  margin-bottom: 1rem;
}

#news-carousel-container .news-carousel-attribution {
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  gap: 1em;
  font-size: 0.85em;
  margin-top: auto;
}

#news-carousel-container .news-carousel-author {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  min-width: 0;
}

#news-carousel-container .news-carousel-date {
  white-space: nowrap;
  flex-shrink: 0;
}
</style>

<script>
document.addEventListener('DOMContentLoaded', function () {
    const carouselContainer = document.getElementById('news-carousel-container');
    const carouselTrack = document.getElementById('news-carousel-track');
    
    if (!carouselContainer || !carouselTrack || !carouselTrack.children.length) {
        return;
    }
    
    const slides = Array.from(carouselTrack.children);
    const displayDuration = 2000;
    let currentTranslate = 0;
    let prevTranslate = 0;
    let currentIndex = 0;
    let intervalId;
    
    // Interaction state variables
    let isDragging = false;
    let startPos = 0;
    let hasDragged = false;
    let wheelTimeout;
    let isWheeling = false;

    const getItemsPerView = () => {
        const width = window.innerWidth;
        if (width <= 768) return 1;
        if (width > 768 && width <= 1024) return 2;
        return 3;
    }

    const getPositionX = (event) => {
        return event.type.includes('mouse') ? event.pageX : event.touches[0].clientX;
    }
    
    const startAutoplay = () => {
        stopAutoplay();
        intervalId = setInterval(autoplayNext, displayDuration);
    }

    const stopAutoplay = () => {
        clearInterval(intervalId);
    }

    const setSliderPosition = () => {
        carouselTrack.style.transform = `translateX(${currentTranslate}px)`;
    }

    const setPositionByIndex = () => {
        if (slides.length === 0) return;

        const itemsPerView = getItemsPerView();
        const maxIndex = slides.length > itemsPerView ? slides.length - itemsPerView : 0;
        
        if (currentIndex > maxIndex) currentIndex = maxIndex;
        if (currentIndex < 0) currentIndex = 0;
        
        const slideWidth = slides[0].getBoundingClientRect().width;
        currentTranslate = currentIndex * -slideWidth;
        
        carouselTrack.style.transition = 'transform 0.4s ease-out';
        setSliderPosition();
    }

    const autoplayNext = () => {
        if (document.hidden || isDragging) return;
        const itemsPerView = getItemsPerView();
        const maxIndex = slides.length > itemsPerView ? slides.length - itemsPerView : 0;
        
        currentIndex++;
        if (currentIndex > maxIndex) {
            currentIndex = 0;
        }
        setPositionByIndex();
    }

    function handleDragStart(event) {
        isDragging = true;
        hasDragged = false;
        startPos = getPositionX(event);
        const style = window.getComputedStyle(carouselTrack);
        const matrix = new DOMMatrix(style.transform);
        prevTranslate = matrix.m41;
        carouselContainer.classList.add('grabbing');
        carouselTrack.style.transition = 'none';
        stopAutoplay();
    }

    function handleDragMove(event) {
        if (!isDragging) return;
        const currentPosition = getPositionX(event);
        currentTranslate = prevTranslate + currentPosition - startPos;
        setSliderPosition();
        if (Math.abs(currentPosition - startPos) > 10) {
            hasDragged = true;
        }
    }

    function handleDragEnd() {
        if (!isDragging) return;
        isDragging = false;
        carouselContainer.classList.remove('grabbing');
        const movedBy = currentTranslate - prevTranslate;
        const itemsPerView = getItemsPerView();
        const maxIndex = slides.length > itemsPerView ? slides.length - itemsPerView : 0;

        if (movedBy < -50 && currentIndex < maxIndex) {
            currentIndex++;
        }
        if (movedBy > 50 && currentIndex > 0) {
            currentIndex--;
        }

        setPositionByIndex();
        startAutoplay();
    }
    
    function handleWheel(event) {
        event.preventDefault();
        if (isWheeling) return;
        isWheeling = true;
        stopAutoplay();

        const itemsPerView = getItemsPerView();
        const maxIndex = slides.length > itemsPerView ? slides.length - itemsPerView : 0;
        const delta = event.deltaY;

        if (delta > 0) {
            if (currentIndex < maxIndex) currentIndex++;
        } else if (delta < 0) {
            if (currentIndex > 0) currentIndex--;
        }
        setPositionByIndex();

        clearTimeout(wheelTimeout);
        wheelTimeout = setTimeout(startAutoplay, 500);
        setTimeout(() => { isWheeling = false; }, 100);
    }
        
    carouselContainer.addEventListener('mousedown', handleDragStart);
    window.addEventListener('mouseup', handleDragEnd);
    window.addEventListener('mousemove', handleDragMove);
    
    carouselContainer.addEventListener('touchstart', handleDragStart, { passive: true });
    window.addEventListener('touchend', handleDragEnd);
    window.addEventListener('touchmove', handleDragMove, { passive: true });

    carouselContainer.addEventListener('click', (e) => {
        if (hasDragged) {
            e.preventDefault();
        }
    }, true);

    carouselContainer.addEventListener('wheel', handleWheel, { passive: false });
    
    carouselContainer.addEventListener('mouseenter', stopAutoplay);
    carouselContainer.addEventListener('mouseleave', startAutoplay);
    document.addEventListener('visibilitychange', () => document.hidden ? stopAutoplay() : startAutoplay());
    
    window.addEventListener('resize', setPositionByIndex);

    setPositionByIndex();
    startAutoplay();
});
</script>
\```
```


</p>
</details>

shravanngoswamii commented:
@mhauru, here is the size and position of the logo: [Preview Link](https://turinglang.org/pr-previews/119#brands). We have a fixed height of 80px, and the width will automatically adjust to maintain the aspect ratio. Please provide the logo in both light mode and dark mode versions. If the dark mode version is unavailable, we can invert the colors for the ATI logo.

Does this work for you? I think it would be helpful for @yebai and @penelopeysm to review the design before we request the logo.

penelopeysm reviewed (commented):
The citation bit is really nice!

simonsteiger commented:
I think the custom page layout (i.e., going all the way to the edges of the viewer) works out for laptop screens but I find my eyes jumping across large distances to capture all the information on a normal size desktop screen:

Laptop:
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/1f4383f1-4401-4371-bd46-e87f6d8e4134" />

Desktop:
<img width="2560" alt="image" src="https://github.com/user-attachments/assets/2a474561-cb09-42ab-b0e5-ec9ffe532bc5" />

I think I'm generally for having some padding on each side, as well as a maximum width for the content to prevent what I described on desktop screens. Were there any advantages you saw for the wide layout over the more narrow one?

shravanngoswamii commented:
> I think I'm generally for having some padding on each side, as well as a maximum width for the content to prevent what I described on desktop screens. What were the advantages you saw for the wide layout over the more narrow one?

Thanks a lot for pointing this out, I actually haven't thought of desktop screen!🥲 You are right, we should have a max width, let me update that too...!

shravanngoswamii commented:
> What were the advantages you saw for the wide layout over the more narrow one?

Actually using custom layout was allowing us to start logo animation header to begin from start of screen to end without any width!

simonsteiger commented:
> Actually using custom layout was allowing us to start logo animation header to begin from start of screen to end without any width!

Oh, that's great! I was sure that there was a good reason for choosing the custom layout, it's just hard to sift through all the changes retrospectively. Thanks you for the explanation!

simonsteiger commented:
> do you have any suggestions for alternate branding option?

Were you thinking about something similar to a logo or a different animation? I don't have any experience creating either of these unfortunately. But if anyone else has those and an idea comes to my mind, I'll certainly share it with you.

The new `max width` layout looks awesome by the way! 

> feel free to take over or push new commits in this PR!

There's currently nothing I'd want to add, but I'm happy to help if you're looking for something specific to be added. One very minor UX point (this might very well be personal taste!) would be that I get confused by elements with hover effects that turn out to be otherwise non-interactive. I'm aware that the hover animation we're using for interactive elements is "louder", but I somehow still perceive it as distracting for the non-interactive elements. Currently, these are the three cards "Expressive", "General-purpose", "Composable", and the cards below "Start Your Journey".

But again, take this with a grain of salt – I absolutely get the visual appeal of hover effects and I'm fine keeping it as is, too. :)

yebai commented on index.qmd:
````diff
@@ -1,55 +1,254 @@
+::: {.d-flex .flex-column .align-items-center .gap-2 .pt-3}
+```{=html}
+<a href="https://turinglang.org/docs/getting-started/" class="button btn">Get Started</a>
+<a href="https://turinglang.org/docs/tutorials/coin-flipping/" class="text-muted" style="font-size: 0.9em;">View Introductory Concepts</a>
````
```suggestion
<a href="https://turinglang.org/docs/tutorials/coin-flipping/" class="text-muted" style="font-size: 0.9em;">Beginner's Walkthrough</a>
```

shravanngoswamii commented:
> Only thing I would change is that in the News section, I'm not sure how the items that are displayed are picked, but currently it shows e.g. a post from 2019 announcing our blog. Maybe it could instead just show the three newest ones?

We just have to put `on_landing: true` in metadata of post to make it appear on landing page!

simonsteiger commented:
Great work @shravanngoswamii !

What do you think about changing the `justify-content: center` property in `.panel-wrapper` to `justify-content: space-between`? This would align the outer borders of the section headers like "What is Turing.jl? [...]" with those of the cards following below.

The aligned version is more appealing to my eyes, but once again this is of course subjective to some degree.

---

With `justify-content: center`:

<img width="1252" height="618" alt="image" src="https://github.com/user-attachments/assets/6e698b56-8394-4210-9eeb-2f1788037ea9" />

---

With `justify-content: space-between`:

<img width="1252" height="618" alt="image" src="https://github.com/user-attachments/assets/1234d491-4673-4895-bd70-60ec77fab715" />

shravanngoswamii commented:
> Maybe there's a way to shrink the card width and force one row until it collapses into the mobile layout with a single column? I'd be happy to tinker with this a bit in the next days; if it doesn't work, we can simply stick with your current solution. :)

Feel free to experiment and push directly to this branch—I’ve added you as a collaborator!

One more favor @simonsteiger, if you have some time: Could you modularize the styles.css and convert all the styles to SCSS, like you did in your original PR? That approach was more optimized than what I’m using now, and I’m not much familiar with SCSS or design principles myself.

simonsteiger commented:
No worries, I'm on it :)

simonsteiger commented:
How much "liberty" do I have regarding the specific grey tones that are currently used in the `styles.css` document? I would try to match the current ones as best as possible, but instead of specifying two or more different hex codes, I'd try get as close as possible with one base colour and then `lighten($color, 5%)` or so.

shravanngoswamii commented:
@simonsteiger Please feel free to change the colors to whatever looks best — we arrived at this grey based on your earlier experiments. Your UI/UX and color sense are far better than mine 😅! We just need to make sure that docs site matches this one: https://github.com/TuringLang/docs/pull/613

shravanngoswamii commented:
> I definitely didn't figure out how to change the hover color of the icons in the navbar, asked Claude and it wasn't of much help either (likely also me not working hard enough to prompt it properly).

Previously, I wasn't using `!important` for links (a:hover {}) so that was being applied to it, now in your updated SCSS, `_layouts.scss` is setting the color for all `.nav-link` elements using an `!important` tag. This prevents the hover effect from applying. We need to add a more specific rule that targets the links on hover and also uses !important to override the base style.

> I felt it was minor enough to leave it for now, but if you @shravanngoswamii recall how you did that, I'd be happy to correct it of course.

I have updated it, no worries!

shravanngoswamii commented:
I have updated docs site PR (https://github.com/TuringLang/docs/pull/613) with SCSS rules that @simonsteiger added in this site!

shravanngoswamii commented:
> The most notable style change I've made is probably to the hover effect of the cards. I found the hover effect of the buttons interesting (grey border by default, no border + translateY + stronger shadow on hover), and applied the same to the hover effects of the cards in both light and dark mode. I like it how it "feels" and also that it's similar effects across the website.

It's looking great so let's stick with it!

shravanngoswamii commented:
> I just checked the docs site PR (https://github.com/TuringLang/docs/pull/613) and felt that the code panel background has suboptimal contrast in dark mode. I guess the procedure of porting the changes between repos is a bit tedious, but let me know @shravanngoswamii if you'd be interested in me having a go at trying to improve these. I didn't have that on my mental map because there's currently no code block on the landing page (or of course News / Team pages).

Hii @simonsteiger, thanks for help! Please feel free to push in both PRs!

penelopeysm commented:
In general I think it looks amazing!

shravanngoswamii commented:
> I didn't have that on my mental map because there's currently no code block on the landing page (or of course News / Team pages).

@simonsteiger, These might help -- https://quarto.org/docs/output-formats/html-themes.html#code-blocks and this:

https://github.com/TuringLang/turinglang.github.io/blob/9f043c67bfe3fa4df46d0fa4ab5092357fe2a2ce/_quarto.yml#L56

simonsteiger commented:
I'll work on the docs PR (https://github.com/TuringLang/docs/pull/613) now and should be done my tomorrow morning at the latest.

I really like the changes to the landing page so far; the new example and the "Get Started" panel left of it looks awesome.
The only very minor change I'd make is to remove the hover effect on the code example because it feels like a different UI element compare to the other cards or buttons. I know that you wanted to keep the hover effects before though, and I'm completely fine with leaving it as is. :)

shravanngoswamii commented:
Thanks so much @simonsteiger for all your help — couldn't have done it this well without you!

shravanngoswamii commented:
Thank you @penelopeysm and @yebai! Let me know if anything else is required to add, I will add them in other PRs, also let's do this someday -- #76

mhauru commented:
The result looks great, thanks @shravanngoswamii!
